db/db_impl/db_impl_open.cc

```diff
@@ -356,6 +356,11 @@ Status DBImpl::Recover(
 
   bool is_new_db = false;
   assert(db_lock_ == nullptr);
+  std::vector<std::string> dbname_children;
+  Status dbname_children_s;
+  if (read_only || immutable_db_options_.best_efforts_recovery) {
+    dbname_children_s = env_->GetChildren(dbname_, &dbname_children);
```

Comment:
I don't think we need a separate `Status dbname_children_s` variable here. You can just define a variable `s` inside the `if` block.

Reply:
This `dbname_children_s` will later be used at least once at the end of the function when `read_only == true`, and may get passed in to TryRecover(). Having `s` in the if block may lose this status.

Reply:
I think we can delay calling `GetChildren()` while still being able to cache. How about the following:
1. best_efforts_recovery is true. Call GetChildren() once when we determine whether to call `NewDB()`. Use the cached result to compute the set of MANIFEST files which we pass to `ManifestPicker()` when calling `TryRecover`. if `read_only` happens to be true, we can also reuse this result when we determine the `options_file_number_`. Although I tend to think `best_efforts_recovery` should be incompatible with read only mode since it will delete SST files. Maybe you can add a PR to disallow the combination of opening a `DBImplReadOnly` with `best_efforts_recovery == true`.
2. best_effort_recovery is false. The first place we call `GetChildren()` should be where we try to find the WAL files. Later, if `read_only && wal_dir == db_name_`, then we can reuse the result.

Reply:
> This `dbname_children_s` will later be used at least once at the end of the function when `read_only == true`, and may get passed in to TryRecover(). Having `s` in the if block may lose this status.

Is it possible just to use a single `Status s` in this function?

Reply:
> I think we can delay calling `GetChildren()` while still being able to cache. How about the following:
> 
> 1. best_efforts_recovery is true. Call GetChildren() once when we determine whether to call `NewDB()`. Use the cached result to compute the set of MANIFEST files which we pass to `ManifestPicker()` when calling `TryRecover`. if `read_only` happens to be true, we can also reuse this result when we determine the `options_file_number_`. Although I tend to think `best_efforts_recovery` should be incompatible with read only mode since it will delete SST files. Maybe you can add a PR to disallow the combination of opening a `DBImplReadOnly` with `best_efforts_recovery == true`.
> 2. best_effort_recovery is false. The first place we call `GetChildren()` should be where we try to find the WAL files. Later, if `read_only && wal_dir == db_name_`, then we can reuse the result.

I'll try to address this in my next commit. I think we may be able to do this, although the code may look a bit messy.

Reply:
> > This `dbname_children_s` will later be used at least once at the end of the function when `read_only == true`, and may get passed in to TryRecover(). Having `s` in the if block may lose this status.
> 
> Is it possible just to use a single `Status s` in this function?

It's tempting, but I don't think so. For example, when `read_only == false`, the first call to `getChildren()` (originally) don't care about returned status and `s` is given pre-programmed values.